efax-gtk-3.2.13/efax/efix.1
.TH EFIX 1 "February 1999" ""  ""
.UC 1
.SH NAME
efix \- convert between fax, text, bit-map and gray-scale formats
.SH SYNOPSIS

.B efix
[
.I options
]
.I file...

.SH OPTIONS

Where \fIoptions\fP are:

.TP 9
.B -i  \fIf\fP
the input image is in format \fIf\fP. Default is to automatically
determine the input type from its contents.

.TP 9
.B 
   fax
fax ("Group3") 1-D coded image

.TP 9
.B 
   text
text.  Line feeds separate lines, form feeds cause page breaks
and tabs are expanded assuming tabs every 8 columns.

.TP 9
.B 
   pbm
raw PBM (portable bit map)

.TP 9
.B 
   tiffg3
TIFF format with Group 3 (fax) compression.

.TP 9
.B 
   tiffraw
TIFF format with no compression.

.TP 9
.B -o  \fIf\fP
write the output in format \fIf\fP.  Default is tiffg3.

.TP 9
.B 
   fax
fax ("Group3") 1-D coded image

.TP 9
.B 
   pbm
raw PBM

.TP 9
.B 
   pgm
raw PGM (Portable Gray Map).  Gray-scale values are produced by
summing pixels in 4x4 pixel blocks.  The output file is 1/4 of
the size given by \-p.  The resulting image has 17 discrete values
between 0 and 255.

.TP 9
.B 
   pcl
HP-PCL (e.g. HP LaserJet).

.TP 9
.B 
   ps
encapsulated Postscript (e.g. Apple Laserwriter).  The file is
compressed using differential coding vertically and run-length
coding horizontally.  There is no provision for positioning the
image within the page and so the image will appear at the lower
left corner of the page when printed.

.TP 9
.B 
   tiffg3
TIFF format with Group 3 (fax) compression.

.TP 9
.B 
   tiffraw
TIFF format with no compression.

.TP 9
.B -n \fIpat\fP
use the printf(3) pattern \fIpath\fP to generate the output file
name.  Up to three %d escapes will be replaced by the page number
starting with 1 (e.g. \-n order.%03d will create file names
order.001, order.002, etc.)

.TP 9
.B -v \fIlvl\fP
print messages of type in string \fIlvl\fP.  Each
\fIlower-case\fP letter in \fIlvl\fP enables one type of message:

.RS 12
.B
e - 
errors
.br
.B
w - 
warnings
.br
.B
i - 
information messages
.br
.B
a - 
program arguments
.br
.B
f - 
file format details
.RE

.RS 9
The default is "ewi".
.RE

.TP 9
.B -f \fIfnt\fP
use font file \fIfnt\fP for text.  The font file for an WxH font
should be a bit map of an image of H rows and 256*W columns.
Each successive WxH cell contains the bit map for characters with
codes from 0 to 255.  The default is to use a built-in 8x16 font.

.TP 9
.B -s \fIX\fP\fRx\fP\fIY\fP
scale the input by a factor of X horizontally and Y vertically.
Scaling does not change the size of the output (use \-p).  If Y is
not specified it is assumed to be the same as X.  Any floating
point value may be used for X and Y. The default is 1,1.

.TP 9
.B -d \fIR\fP\fR,\fP\fID\fP
displace the output right by R and down by D (opposite if
negative). See below for units.  Default is 0,0.

.TP 9
.B -p \fIW\fP\fRx\fP\fIH\fP
truncate or pad the output to generate an image of width W and
height H.  This does not scale the input.  See below for units.
The default is the size of the input image if it can be
determined or A4 (215x297mm) if it can't.

.TP 9
.B -r \fIX\fP\fRx\fP\fIY\fP
assume an output device resolution of X by Y dots per inch.  If Y
is not specified it is assumed to be the same as X.  The default
is the input resolution if it can be determined or the fax
resolution of 204.1x195.6 dpi if it can't.

.TP 9
.B -R \fIX\fP\fRx\fP\fIY\fP
assume an input device resolution of X by Y dots per inch.  If Y
is not specified it is assumed to be the same as X.  The default
is the input resolution if it can be determined or the fax
resolution of 204.1x195.6 dpi if it can't.

.TP 9
.B -l \fIn\fP
place n lines per page during text input. Default is 66.

.TP 9
.B -O \fIf\fP
overlay (logical OR) the image from file f into the output.  Use
"\-" for standard input (\-O\-).  Default is no overlay file.

.TP 9
.B -M
ignore all other options and copy the standard input to the
standard output while applying base64 (MIME) encoding as
specified by RFC 1521.


.SH FILES

If no \-n options are given, output is written to the standard
output.

.SH UNITS

The units of the W, H, R, and D values above are in inches by
default.  Any floating point value may be used.  Units of inches,
centimetres, millimetres or points (72 per inch) can be used
instead by appending one of the strings `in', `cm', `mm', or `pt'
to the argument (e.g. \-d2,4cm).

.SH CUT AND PASTE

The \-d and \-p options allow efix to cut out images from received
faxes for use in other faxes or documents.  The \-d option specifies
the top left portion of the desired image and the \-p option gives
the size of the cut image.  For example, the command
.RS
.nf
.ft CW
	efix \-d-5,\-8 \-p2,1 sample.001 >sig.001
.ft P
.fi
.RE
would cut out part of the input with its top left corner 5 inches
from the left edge and 8 inches from top of the input image.  The
output image would be 2 inches wide and 1 inch high.

The \-O option allows efix to superimpose two or more images.  The
overlay image must be in fax format and cannot be scaled,
truncated or shifted. However, multiple efix commands may be used
to transform images before combining them.  For example, the
commands
.RS
.nf
.ft CW
	efix \-d4,8 signature >sig.fax
	efix \-O sig.fax letterhead >letterhead.fax
	efix \-O letterhead.fax letter.002 >letter.002.new
.ft P
.fi
.RE
will shift the image in the file signature down 8 inches and
right 4 inches and combine (overlay) it with the images in the
files letterhead and letter.002.

.SH REFERENCES

Gunter Born, "The File Formats Handbook", International Thompson
Computer Press, 1995.

.SH COPYRIGHT

efix is copyright 1994 -- 1999 by Ed Casas.  It may be used,
copied and modified under the terms of the GNU Public License.

.SH DISCLAIMER

Although \fBefix\fP has been tested it may have errors that will
prevent it from working correctly on your system.  Some of these
errors may cause serious problems including loss of data.

.SH SEE ALSO

.BR efax(1),
.BR ghostscript(1),
.BR pbm(5), 
.BR pgm(5).

.SH  BUGS

Only reads two types of TIFF compression formats.

Does not write multi-page TIFF files (a feature).



iaxmodem-1.2.0/CHANGES
1.2.0
	improve skew behavior, especially for faulty RTC; add "skew" config feature
	add support for command-dial (ATD...;) in spandsp
	add support for ATA for CED when off-hook and connected to a call
	add support for ATD for CNG when off-hook and connected to a call
	fix the ordering of the seteuid()/setegid() (Julien BLACHE)
	update spandsp to 20080725 snapshot

1.1.1
	fix unaligned errors in libiax2 (Gus Bourg)
	improve training in spandsp V.27ter receive (Steve Underwood)
	fix a potential crash (divide by zero) in spandsp V.17 receive (Steve Underwood)
1.1.0
	kill -HUP now makes the modems wait to restart when they're on-hook
	improve IAX2 call rejection when the modem is busy by using
	    REJECT instead of ACCEPT+CONGSTN+HANGUP
	add "nodaemon" config file feature
	add "iax2debug" and "dspdebug" config file features
	update spandsp to 20080110 snapshot
	update libiax2 to current iaxclient/lib/libiax2 + patches
1.0.0
	provide strndup for Solaris
0.3.2
	support dialing of DTMF and commas
	support Solaris (thanks, Bruce McAlister)
	fix receive aborts when carrier is connected
	fix +FRH:3 from appearing at improper times
	set hang-up timer on unexpected disconnections as well
	stop fast looping on registration rejection
	fix numerous training issues with V.17 rx
	update spandsp to snapshot 20070802
0.3.1
	add -F option to prevent detachment (Francesco P. Sileno)
	update spandsp to snapshot 20070619
	fix possible initial failure to register
0.3.0
	skip non-regular files in /etc/iaxmodem (Julien BLACHE)
	take caution against bad refresh negotiations from the server
	update spandsp to snapshot 20070502 (adds V.17 rx support)
	improve the ability to detect quiet signalling (spandsp: fsk min_power)
	improve the sensitivity to carrier loss (spandsp: power_meter)
	improve tcflushing when getty not reading pty and buffer fills,
	    prevents fast looping when buffer is filled
0.2.1
	update spandsp to snapshot 20070123
	on startup rotate log files larger than 1GB, open with O_LARGEFILE
	use power level of detected signalling in silence identification
0.2.0
	add CDID, USER, and PASS (context, username, password) in
	    call identification information
	add feature-usage of posix_openpt (Julien BLACHE)
	fix the error of CONNECT appearing when in command-mode (spandsp)
	update spandsp to snapshot 20061217
	add nojitterbuffer configuration option
	fix jitterbuffer to send non-null audio
	fix some array size off-by-one counting problems (Hans Petter Selasky)
0.1.16
	update spandsp to snapshot 20061116
	fix some select racing due to invalid selects and just before NO ANSWERs
	add "restart modem" feature (kill -HUP) for individual modems
	fix problem of sending DTE data after a final-frame HDLC
	    message but before the OK and carrier drop
	add jitterbuffer
0.1.15
	attempt to compensate for received jitter
	rename recorded files to "old" when done recording
	update spandsp to snapshot 20060907 (and fix spandsp)
	improve audio file closing
	add configure and Makefile.in (contrib: Anish Mistry)
	fix library versioning for static builds
0.1.14
	take measures to avoid detecting echo as signalling
	ignore spandsp "on hook" handler messages
	ignore DTMF when not received in command-mode
	more finely tune V.29 and V.27ter rx training (spandsp)
0.1.13
	properly terminate Phase C reception data when an unexpected
	    disconnection occurs there
	fix idle-busy condition (report: Francesco Romano, Olaf Ferger)
	flush pty buffer if it fills
0.1.12
	cause the AT command interpreter to be more strict (spandsp)
	fix condition where DTE input would not be passed on to spandsp
	    when it should have been
	enable data flow if flow control is off on hangup (spandsp)
	add IAX2 DTMF recognition and reporting
	fix V.21 HDLC flags threshold detection (spandsp)
0.1.11
	update spandsp to 0.0.3-snapshot-20060707
	add support for comments begun by ";" in the config files
	fix "Setting refresh" output information
0.1.10
	add 5-second automatic busy-out after hangups
	ensure use of iax_destroy(session) after iax_hangups
	removed -O2 from build script as reports claim it helps stability
	implement an inactivity timeout in stuffed tx mode (spandsp)
	fix -V version output for libiax2 and spandsp
0.1.9
	configure and make libs automatically with 'build static'
	apply fix for CORE-2006-0327 libiax2 vulnerability
	properly place the modem on-hook on a hangup (spandsp)
	don't skew due to incoming audio before the call is answered
0.1.8
	log received IAX2 jitter
	fix a logic error that could cause fast looping (report Lee Archer)
	fix "no room in buffer" error messages
	fix IAX2 timestamp interpretation on miniframes (libiax2)
	fix md5 auth on x86-64 (from iaxclient, Steve Kann, Steve Underwood)
	add quoting on version defines (Stefan Knoblich)
0.1.7
	update spandsp to 0.0.3-snapshot-20060525 (fixes CNG, report Xavier Caro)
0.1.6
	add support for AT+VSID="%s","%d", Caller*ID setting from DTE
	improve V.17 rx training (spandsp, Bartek Kania, Steve Underwood)
	update spandsp to 0.0.3-snapshot-20060523
0.1.5
	fix x86-64 spandsp build problem (report: Julien BLACHE, 
	    fix: Steve Underwood)
	re-fix potentially missing V.21 CONNECT messages (spandsp)
0.1.4
	handle error situation where DSP may not have any available
	    space for more tty data (report: Martin Vit)
	adjust iaxmodem.c and libiax2 for OpenBSD (Todd T. Fries)
	update spandsp to 0.0.3-snapshot-20060427
	add command-line -V (version) argument and response
0.1.3
	fix potentially missing V.21 CONNECT messages (spandsp)
	ignore spandsp's ring indicator
	add auto-answer (e.g. ATS0=1) support (spandsp)
	make V.29 rx more sensitive (spandsp)
0.1.2
	improve logging features (Julien BLACHE)
	cope with audio that is non-zero centered (spandsp, Steve Underwood)
	fix lrintf() warning on static builds (Julien BLACHE)
0.1.1
	fix timeout on calling for initial carrier detection (spandsp)
	add silent dial support (spandsp)
	don't loop fast when not registering

efax-gtk-3.2.13/efax/efix.c
#define Copyright         "Copyright 1999 Ed Casas"

#define Version		  "efix v 0.3"

/*
    Copyright (C) 1999  Ed Casas

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

    Please contact the author if you wish to use efax or efix in
    ways not covered by the GNU GPL.

    You may contact the author by e-mail at: edc@cce.com.

*/

/*
    The call to setlocale(LC_NUMERIC, "C") was added by Chris Vine 13
    September 2005 to avoid incorrect floating point representation in
    locales which use the comma instead of the full-stop as the
    decimal separator.

    The following patche from http://shino.pos.to/linux/efax/ also
    applied:

      efax-0.9-numlines.patch
*/

const char *Usage =
  "Usage:\n"
  "  %s [ option ]... file... \n"
"Options (defaults):\n"
  "  -i  f   input format (auto):\n"
  "     fax     fax (\"Group3\") 1-D coded image\n"
  "     text    text\n"
  "     pbm     raw PBM (portable bit map)\n"
  "     tiffg3  TIFF, Group 3 fax compression\n"
  "     tiffraw TIFF, no compression\n"
  "     pcx     mono PCX\n"
  "     dcx     mono DCX\n"
  "  -o  f   output format (tiffg3):\n"
  "     fax     fax (\"Group3\") 1-D coded image\n"
  "     pbm     Portable Bit Map\n"
  "     pgm     Portable Gray Map (decimated by 4)\n"
  "     pcl     HP-PCL (e.g. HP LaserJet)\n"
  "     ps      Postscript (e.g. Apple Laserwriter)\n"
  "     tiffg3  TIFF, Group 3 fax compression\n"
  "     tiffraw TIFF, no compression\n"
  "     pcx     mono PCX\n"
  "     dcx     mono DCX\n"
  "  -n pat  printf() pattern for output file name (ofile)\n"
  "  -f fnt  use PBM font file fnt for text (built-in)\n"
  "  -l  n   lines per text page (66)\n"
  "  -v lvl  print messages of type in string lvl (ewi)\n"
  "  -s XxY  scale input by X and Y (Y optional) (1x1)\n"
  "  -r XxY  resolution of output is X by Y (dpi, Y optional) (204x196)\n"
  "  -R XxY  resolution of input  is X by Y (dpi, Y optional) (204x196)\n"
  "  -p WxH  pad/truncate output to width W by height H (215x297mm)\n"
  "  -d R,D  displace output right R, down D (opposite if -ve) (0,0)\n"
  "  -O f    overlay file f (none)\n"
  "  -M      ignore other options and base64 (MIME) encode stdin to stdout\n"
  "\n"
  "Add 'in', 'cm', 'mm', or 'pt' to -p and -d arguments (default in[ches]).\n" 
  "Default output size and resolution is same as input (if known).\n" 
  ;

#include <ctype.h>		/* ANSI C */
#include <limits.h>
#include <locale.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "efaxlib.h"
#include "efaxmsg.h"

#ifndef INT_MAX
#define INT_MAX 32767
#endif

/* Allowed input and output formats. *** MUST match enum *** */

char *iformatstr[] = { " 3text", " 1pbm", " 2fax", " 4tiffg3", " 4tiffraw", 
		       " 6pcx", " 6pcxraw", " 8dcx", 0 } ;

char *oformatstr[] = { " 1pbm" , " 2fax", " 3pcl", " 4ps",  " 5pgm", 
		       " 7tiffg3", " 8tiffraw", 
		       "11pcx", "12pcxraw", "13dcx", 0 } ;

/* Look up a string in a NULL-delimited table where the first
   character of each string is the digit to return if the rest of
   the string matches.  Returns the value of the digit for the
   matching string or -1 if no matches. */

int lookup ( char **tab, char *s )
{
  char **p ;
  for ( p=tab ; p && *p && strcmp ( *p+2, s ) ; p++ ) ;
  return p && *p ? atoi ( *p ) : -1 ;
}


/* Extract pair of values from string.  If it's a `dim'ension,
   two values are required and they are converted to inches, else
   the y value is optional.  Returns 0 or 2 on error. */

int getxy ( char *arg, float *x, float *y, int dim )
{
  int i, n, nc=0, err=0 ;
  char c ;
  static char *unitstr[] = { " 0in", " 1cm", " 2mm", " 3pt", 0 } ;
  static float unitval[] = { 1.0, 2.54, 25.4, 72.0, 1.0 } ;

  if ( ! arg ) 
    err = msg ( "E2 missing argument" ) ;

  if ( !x || !y )
    err = msg ( "E2 can't happen (getxy)" ) ;

  if ( ! err ) {
    n = sscanf ( arg , "%f%c%f%n", x, &c, y, &nc ) ;
    switch ( n ) {
    case 0 : err = msg ( "E2bad X value in (%s)", arg ) ; break ;
    case 2 : err = msg ( "E2bad Y value in (%s)", arg ) ; break ;
    }      
  }

  if ( ! err ) {
    if ( dim ) {
      if ( n != 3 ) {
	err = msg ( "Emissing Y dimension in (%s)", arg ) ;
      } else {
	while ( arg [ nc ] && isspace ( arg [ nc ] ) ) nc++ ;
	if ( arg [ nc ] ) {
	  if ( ( i = lookup ( unitstr, arg+nc ) ) >= 0 ) {
	    *x /= unitval [ i ] ;
	    *y /= unitval [ i ] ;
	  } else {
	    err = msg ( "E2bad units: `%s'", arg+nc ) ;
	  }
	}
      }
    } else {
      if ( n == 1 ) *y = *x ;
    }
  }
  
  if ( ! err )
    msg ( "Aconverted (%s) into %f x %f", arg, *x, *y ) ;

  return err ;
}


/* Copy stdin to stdout while applying base64 (RFC 1521)
   encoding.  This encoding must be applied after the file is
   complete since some output formats (e.g. TIFF) require seeking
   backwards within the (binary) file.  */

int base64encode ( void )
{
  int err=0, c ;
  uchar n=0, m=0, bits=0 ;

  static uchar chartab[65] = 
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ" 
    "abcdefghijklmnopqrstuvwxyz"
    "0123456789+/" ;

  while ( ( c = fgetc ( stdin ) ) >= 0 ) {
    switch ( n ) {
    case 0:
      putc ( chartab [ c >> 2 ], stdout ) ;
      bits = c & 0x3 ;
      n = 1 ;
      break ;
    case 1:
      putc ( chartab [ (bits << 4) | ( c >> 4 ) ], stdout ) ;
      bits = c & 0xf ;
      n = 2 ;
      break ;
    case 2:
      putc ( chartab [ (bits << 2) | ( c >> 6 ) ], stdout ) ;
      putc ( chartab [ c & 0x3f ], stdout ) ;
      n = 0 ;
      if ( ++m >= 18 ) {
	putc ( '\n', stdout ) ;
	m = 0 ;
      }
      break ;
    }

  }

  switch ( n ) {
  case 0:
    break ;
  case 1:
    putc ( chartab [ (bits << 4) | ( 0 >> 4 ) ], stdout ) ;
    putc ( '=', stdout ) ;
    putc ( '=', stdout ) ;
    break ;
  case 2 :
    putc ( chartab [ (bits << 2) | ( 0 >> 6 ) ], stdout ) ;
    putc ( '=', stdout ) ;
    break ;
  }

  putc ( '\n', stdout ) ;

  return err ;
}


int main( int argc, char **argv)
{
  int err=0, done=0, i, c ;
  int nr, pels, ovnr, ovpels, no ;	/* run/pixel/repeat counts */
  int linesout ;
  int page, ilines, olines ;		/* page & line counts */
  int xs, ys, w, h, ixsh, iysh ;	/* integer scale, size & shift */
  short runs [ MAXRUNS ] , ovruns [ MAXRUNS ] ;
  
  float					 /* defaults: */
    xsc=1.0, ysc=1.0,		         /* scale */
    xsh=0.0, ysh=0.0,			 /* shift */
    dxres = 204.145,			 /* o/p res'n: 1728/215mm * 25.4 x */
    dyres = 195.58,			 /* 7.7 * 25.4 */
    dxsz = 215 / 25.4,			 /* o/p size: 8.5" x A4 */
    dysz = 297 / 25.4 ;

  float				/* arguments: */
    axres = 0, ayres = 0, axsz = 0, aysz = 0, ainxres=0, ainyres=0 ;

  float				/* values used: */
    xres = 0, yres = 0, xsz = 0, ysz = 0 ;

  IFILE ifile, ovfile ;
  OFILE ofile ;

  char **ifnames,  *ovfnames [ 2 ] = { 0, 0 } ;

  int iformat=I_AUTO, oformat=O_TIFF_FAX, pglines=0 ;
  char *ofname=0 ;

  faxfont font, *pfont=0 ;	/* text font */

  /* initialize */

  argv0 = argv[0] ;

  setlocale ( LC_ALL, "" ) ;
  /*
    efix uses formatted text functions for floating point numbers, so restore
     the C locale for that
  */
  setlocale ( LC_NUMERIC, "C" ) ;

  /* process arguments */

  while ( !err && (c=nextopt(argc,argv,"n:i:o:O:v:l:f:r:s:p:d:R:M") ) != -1) {
    switch ( c ) {
    case 'n':
      ofname = nxtoptarg ;
      break ;
    case 'i': 
      if ( ( iformat = lookup ( iformatstr, nxtoptarg ) ) < 0 ) 
	err = msg ( "E2invalid input type (%s)", nxtoptarg ) ;
      break ;
    case 'o': 
      if ( ( oformat = lookup ( oformatstr, nxtoptarg ) ) < 0 )
	err = msg ( "E2invalid output type (%s)", nxtoptarg ) ;
      break ;
    case 'O': 
      ovfnames[0] = nxtoptarg ;
      break ;
    case 'v': 
      verb[0] = nxtoptarg ;
      msg ( "A " Version ) ;
      for ( i=0 ; i<argc ; i++ ) msg ( "Aargv[%d]=%s", i, argv[i]) ;
      break ;
    case 'l':
      if ( sscanf ( nxtoptarg , "%d", &pglines ) != 1 || pglines <= 0 ) {
	err = msg ( "E2bad page length (%s)", nxtoptarg ) ;
	pglines = 0 ;
      }
      break ;
    case 'f' :
      if ( ! ( err = readfont ( nxtoptarg, &font ) ) )
	pfont = &font ;
      break ;
    case 's' : err = getxy ( nxtoptarg, &xsc , &ysc , 0 ) ; break ;
    case 'r' : err = getxy ( nxtoptarg, &axres, &ayres, 0 ) ; break ;
    case 'R' : err = getxy ( nxtoptarg, &ainxres, &ainyres, 0 ) ; break ;
    case 'p' : err = getxy ( nxtoptarg, &axsz , &aysz , 1 ) ; break ;
    case 'd' : err = getxy ( nxtoptarg, &xsh , &ysh , 1 ) ; break ;
    case 'M' : err = base64encode() ; done=1 ; break ;
    default : fprintf ( stderr, Usage, argv0 ) ; err = 2 ; break ;
    }
  }

  msg ( "I " Version " " Copyright ) ;

  if ( ! err && ! done ) {

    if ( pfont ) ifile.font = pfont ;
    if ( pglines ) ifile.pglines = pglines ;

    if ( nxtoptind < argc ) {
      ifnames = argv + nxtoptind ;
      if ( argv [ argc ] ) {
	err = msg ("E2can't happen(unterminated argv)") ;
      } else {
	newIFILE ( &ifile, ifnames ) ;
      }
    } else {
      err = msg ( "E3 missing input file name" ) ;
    }

    newIFILE ( &ovfile, ovfnames ) ;

    newOFILE ( &ofile, oformat, ofname, 0, 0, 0, 0 ) ;

  }

  for ( page = 0 ; ! err && ! done ; page++ ) {

    if ( nextipage ( &ifile, page != 0 ) ) { 
      done=1 ; 
      continue ; 
    }

    /* set output size and resolution equal to input if none specified */

    if ( ainxres > 0 ) ifile.page->xres = ainxres ;
    if ( ainyres > 0 ) ifile.page->yres = ainyres ;

    if ( ifile.page->xres <= 0 ) ifile.page->xres = dxres ;
    if ( ifile.page->yres <= 0 ) ifile.page->yres = dyres ;

    xres = axres > 0 ? axres : ifile.page->xres ;
    yres = ayres > 0 ? ayres : ifile.page->yres ;

    xsz = axsz > 0 ? axsz : ( ifile.page->w > 0 ? 
			      ifile.page->w / ifile.page->xres : dxsz ) ;
    ysz = aysz > 0 ? aysz : ( ifile.page->h > 0 ? 
			      ifile.page->h / ifile.page->yres : dysz ) ;


    w = xsz * xres + 0.5 ;	      /* output dimensions in pixels */
    h = ysz * yres + 0.5 ;
    
    ixsh = xsh * xres ;		      /* x/y shifts in pixels/lines */
    iysh = ysh * yres ;
    
    if ( ( w & 7 ) != 0 )	/* just about everything requires... */
      msg ("Iimage width rounded to %d pixels", 
	   w = ( w + 7 ) & ~7 ) ;
    
    if ( ofile.format == O_PGM && h & 3 ) /* PGM x4 decimation requires... */
      msg ("I PGM image height rounded up to %d lines", 
	   h = ( h + 3 ) & ~3 ) ;
    
    if ( w <= 0 || h <= 0 || xres < 0 || yres < 0 )
      err = msg ( "E2negative/zero scaling/size/resolution" ) ;
    
    if ( ofile.format == O_PCL &&	/* check for strange PCL resolutions */
	( xres != yres || ( xres != 300 && xres != 150 && xres != 75 ) ) )
      msg ( "Wstrange PCL resolution (%.0fx%.0f)", xres, yres ) ;
    
    if ( w > MAXBITS*8 )	/* make sure output will fit... */
      err = msg( "E2requested output width too large (%d pixels)", w ) ;
    
    ofile.w = w ; 
    ofile.h = h ; 
    ofile.xres = xres ; 
    ofile.yres = yres ;

    /* scale according to input file resolution */

    xs = 256 * xsc * xres / ifile.page->xres + 0.5 ;
    ys = 256 * ysc * yres / ifile.page->yres + 0.5 ;

    if ( xs <= 0 || ys <= 0 )
      err = msg ( "E2negative/zero scaling" ) ;

    if ( *ovfnames )		      /* [re-]open overlay file */
      if ( nextipage ( &ovfile , 0 ) ) { 
	err=2 ; 
	continue ; 
      }

    if ( nextopage ( &ofile, page ) ) { 
      err=2 ; 
      continue ; 
    }
    linesout=0 ;

    /* y-shift */

    if ( iysh > 0 ) {
      writeline ( &ofile, ( ( *runs = w ), runs ), 1, iysh ) ;
      linesout += iysh ;
    } else {
      for ( i=0 ; i < -iysh ; i++ ) 
	readline ( &ifile, runs, 0 ) ;
    }    

    /* copy input to output */
    
    olines = ilines = 0 ; 
    
    while ( linesout < h ) {

      if ( ! ifile.lines || ( nr = readline ( &ifile, runs, &pels ) ) < 0 ) {
	break ;
      } else {
	ilines++ ;
      }

      if ( *ovfnames ) {
	if ( ( ovnr = readline ( &ovfile, ovruns, &ovpels ) ) >= 0 )
	  nr = runor ( runs, nr, ovruns, ovnr, 0, &pels ) ; 
      }

      /* x-scale, x-shift & x-pad input line */
    
      pels  = ( xs == 256 ) ? pels : xscale ( runs, nr, xs ) ;
      pels += ( ixsh == 0 ) ?   0  : xshift ( runs, nr, ixsh ) ;
      nr    = ( pels == w ) ?  nr  : xpad   ( runs, nr, w - pels ) ;

      /* y-scale by deleting/duplicating lines. */

      no = ( ( ilines * ys ) >> 8 ) - olines ;

      if ( linesout + no > h ) no = h - linesout ;
      olines += no ;

      writeline ( &ofile, runs, nr, no ) ;
      linesout += no ;
    }

    /* y-pad */

    if ( linesout < h )
      writeline ( &ofile, ( ( *runs = w ), runs ), 1, h - linesout ) ;
    
    if ( ferror ( ifile.f ) ) err = msg ( "ES2input error:" ) ;
  }

  nextopage ( &ofile, EOF ) ;

  return err ;
}
